aiida/tools/visualization/graph.py

```diff
@@ -0,0 +1,740 @@
+                 data_styles=None,
+                 process_styles=None,
+                 data_sublabels=None):
+        """a class to create graphviz graphs of the AiiDA node provenance
+
+        Nodes and edges, are cached, so that they are only created once
+
+        :param engine: the graphviz engine, e.g. dot, circo (Default value = None)
+        :type engine: str or None
+        :param graph_attr: attributes for the graphviz graph (Default value = None)
+        :type graph_attr: dict or None
+        :param global_node_style: styles which will be added to all nodes.
+            Note this will override any builtin attributes (Default value = None)
+        :type global_node_style: dict or None
+        :param global_edge_style: styles which will be added to all edges.
+            Note this will override any builtin attributes (Default value = None)
+        :type global_edge_style: dict or None
+        :param include_sublabels: if True, the note text will include node dependant sub-labels (Default value = True)
+        :type include_sublabels: bool
+        :param link_styles: callable mapping LinkType to graphviz style dict;
+            link_styles(link_type) -> dict (Default value = None)
+        :param data_styles: callable mapping data node to a graphviz style dict;
+            data_styles(node) -> dict (Default value = None)
+        :param process_styles: callable mapping process node to a graphviz style dict;
+            process_styles(node) -> dict (Default value = None)
+        :param data_sublabels: callable mapping data node to a sublabel (e.g. specifying some attribute values)
+            data_sublabels(node) -> str (Default value = None)
+
+        """
+        # pylint: disable=too-many-arguments
+        self._graph = Digraph(engine=engine, graph_attr=graph_attr)
+        self._nodes = set()
+        self._edges = set()
+        self._global_node_style = {}
+        if global_node_style:
+            self._global_node_style = global_node_style
+        self._global_edge_style = {}
+        if global_edge_style:
+            self._global_edge_style = global_edge_style
+        self._include_sublabels = include_sublabels
+        if link_styles is not None:
+            self._link_styles = link_styles
+        else:
+            self._link_styles = default_link_styles
+        if data_styles is not None:
+            self._data_styles = data_styles
+        else:
+            self._data_styles = default_data_styles
+        if process_styles is not None:
+            self._process_styles = process_styles
+        else:
+            self._process_styles = default_process_styles
+        if data_sublabels is not None:
+            self._data_sublabels = data_sublabels
+        else:
+            self._data_sublabels = default_data_sublabels
+
+    @property
+    def graphviz(self):
+        """return a copy of the graphviz.Digraph"""
+        return self._graph.copy()
+
+    @property
+    def nodes(self):
```

Comment:
I guess this copying is also actually needed where `.nodes` is used?
I.e. there is no excessive copying going on?